### ISSUE TITLE
need documentation on configuring minikube with gitlab

### ISSUE THREAD

**Issue by Divine1** [eclipse/che#21061]
### Is your task related to a problem? Please describe

there is no documentation that shows how to configure minikube with gitlab.

i can see the documentation for configuring minikube with github (https://www.eclipse.org/che/docs/che-7/administration-guide/configuring-minikube-github-authentication/)

### Describe the solution you'd like

if there is documentation on how to configure minikube with gitlab, it will be helpful for me.

### Describe alternatives you've considered
i haven't considered any other alternatives.
i'm amateur in this area, so i'm looking for documentation.

**Comment by svor** [comment]
@Divine1 i think the documentation you are looking for will be updated/added by https://github.com/eclipse/che-docs/pull/2202

**Comment by skabashnyuk** [comment]
If I understand correctly @Divine1, he want's an example of how to set up `dex` for GitLab authentication. So the whole k8s cluster authentication would be connected to Gitlab. 

@Divine1 did you already try that? Did you have any issues to setup Gitlab in this role?

**Comment by Divine1** [comment]
@skabashnyuk yes you understood correctly.

currently im checking if `github` oauth integration is working fine.  Once its done i thought of using below configuration for `gitlab` by editing the configMap and updating below information `kubectl edit configmap dex -n dex`



```
connectors:
- type: gitlab
  id: gitlab
  name: GitLab
  config:
    clientID: <client_id> 
    clientSecret: <client_secret> 
    redirectURI: https://dex.<minikube_ip>.nip.io/callback 
```

**Comment by skabashnyuk** [comment]
@Divine1 ok. And what do you get as a result? Is there an error somewhere?

**Comment by skabashnyuk** [comment]
connectors are looks like on the same level with the `data` field of configMap. I believe it should be inside `config.yaml`.

**Comment by Divine1** [comment]
@skabashnyuk 


yes , you are right. i updated the configmap as per your suggestion, it the gitlab oauth authentication was successful.

below is the config file which i used for github auth integration with dex. Now i'm going to try oauth integration with gitlab by following the same detail for github.
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
data:
  config.yaml: |
    issuer: https://dex.192.168.49.2.nip.io
    storage:
      type: kubernetes
      config:
        inCluster: true
    web:
      http: 0.0.0.0:5556

    oauth2:
      skipApprovalScreen: true

    staticClients:
    - id: eclipse-che
      redirectURIs:
      - 'https://192.168.49.2.nip.io/oauth/callback'
      name: 'Eclipse Che'
      secret: R5ysT5ONfIzHBpgl8UrFBxOdJamsELe3secret=
    connectors:
    - type: github
      id: github
      name: GitHub
      config:
        clientID: 405fc0d53034dclient
        clientSecret: c23230d0b495f5c5a0054cc0c648esecret
        redirectURI: https://dex.192.168.49.2.nip.io/callback
kind: ConfigMap
metadata:
  creationTimestamp: "2022-01-18T05:03:41Z"
  labels:
    app: dex
  name: dex
  namespace: dex
  resourceVersion: "2284745"
  uid: 5156ba91-88d9-4115-b9ad-c07b0b5195c1

```

**Comment by skabashnyuk** [comment]
FYI. gitlab for devworkspaes is not yet merged. See more https://github.com/eclipse-che/che-server/pull/195

**Comment by Divine1** [comment]
@skabashnyuk  is there any solution available which i can follow now?

**Comment by skabashnyuk** [comment]
you can build custom che-server image from https://github.com/eclipse-che/che-server/pull/195

**Comment by Divine1** [comment]
@skabashnyuk  thanks a lot for this info.

What steps should i follow to build a custom che-server image from [eclipse-che/che-server#195](https://github.com/eclipse-che/che-server/pull/195) ?

**Comment by Divine1** [comment]
@skabashnyuk  thank you.

after step5 what should i do?

**Comment by Divine1** [comment]
@skabashnyuk  i'm not able to checkout the branch you have shared. is that branch merged and deleted?

**Comment by Divine1** [comment]
@skabashnyuk 

i cloned the `main` branch from [eclipse-che/che-server](https://github.com/eclipse-che/che-server.git)  
then i executed the `mvn install` command and built the image

the installed the image using below command. 
`chectl server:deploy --cheimage=quay.io/eclipse/che-server:next --platform=minikube`

but i received the `keycloak url` in the installation logs instead of `dex` related configuration. 

i'm not sure whether the image which i built got installed here or the image from central repository got pulled.

please let me know your feedback.
![image](https://user-images.githubusercontent.com/10482177/151030106-44603b27-181c-422e-b9f3-52e1f0f323ee.png)

**Comment by skabashnyuk** [comment]
@Divine1 FYI. I've just merged https://github.com/eclipse-che/che-server/pull/195. Soon it has to be built by our CI.

**Comment by Divine1** [comment]
@skabashnyuk thank you for this update. 
in which version of the software will this fix be available? 
once this CI build is completed will i be able to install via `chectl`  command?
is it released now?

**Comment by skabashnyuk** [comment]
7.43

**Comment by Divine1** [comment]
@skabashnyuk how do i install `7.43` using `chectl`?

**Comment by skabashnyuk** [comment]
`chectl update next` ?

**Comment by Divine1** [comment]
i did a fresh installation of eclipse-che on minikube. i receive below error now. how to resolve this error?

ubuntu 18.04 - 64bit 

`minikube start --driver=docker --memory=8192`

`chectl server:deploy --platform minikube --installer=operator`

![image](https://user-images.githubusercontent.com/10482177/151317529-9559f0d4-4b47-4115-8714-3f30660836fa.png)


this is the chectl version that i use currently
![image](https://user-images.githubusercontent.com/10482177/151317720-7eba21af-63bf-4be1-87e8-8d31a538ee1b.png)

**Comment by Divine1** [comment]
@skabashnyuk  the installation is successful now. i deleted the minikube cluster, then followed below steps. i inferred this by looking around for solutions for the error's i posted in the previous comment.

1. `minikube start --driver=docker --memory=8192` (leave out ingress add-on and do it manually later)
2. `chectl server:deploy --platform minikube --installer=operator --debug` (this command will not succeed)
3.  `minikube addons enable ingress` (now enable ingress on minikube)
4.  `chectl server:deploy --platform minikube --installer=operator --debug` (re-run the deploy command)
5. installation success now but i feel like this problem is due to internet slowness, but i'm not able to prove it.




now i have to integerate`oauth-gitlab` with `eclipse-che` Similar to how i integrated oauth-github with `eclipse-che` by following [this link](https://www.eclipse.org/che/docs/che-7/administration-guide/configuring-minikube-github-authentication/)


i have taken below yaml from [this link](https://github.com/eclipse-che/che-docs/pull/2202/files#diff-321a9664a608252ec41923967906b1a3e605bc576043ea0d1bab03060407df4bR19) which you have shared above to integerate`oauth-gitlab` with `eclipse-che`.

```
kind: Secret
apiVersion: v1
metadata:
  name: gitlab-oauth-config
  namespace: __<namespace>__ <1>
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: oauth-scm-configuration
  annotations:
    che.eclipse.org/oauth-scm-server: gitlab
    che.eclipse.org/scm-server-endpoint: __<your_GitLab_service_URL>__
type: Opaque
data:
  id: __<base64_encoded_GitLab_Application_ID>__
  secret: __<base64_encoded_GitLab_Secret>__


```

**under which namespace should i apply this yaml ?**

![image](https://user-images.githubusercontent.com/10482177/151328218-e87f9d20-5bf1-4e04-9ab7-dfffdb1f6f76.png)

**Comment by skabashnyuk** [comment]
```eclipse-che```

**Comment by Divine1** [comment]
@skabashnyuk 

i have to remove Dex static users, delete all enablePasswordDB and staticPasswords sections from `kubectl edit configmap dex -n dex` configMap inorder to make the `oauth-gitlab-integration` successful, right?



![image](https://user-images.githubusercontent.com/10482177/151336756-b205d1da-5cea-4444-a0de-29f3213b5600.png)

**Comment by skabashnyuk** [comment]
There are two distinct topics. Gitlab as oidc provider for Kubernetes (eclipse che) authentication and the second one authentification of `git clone/push` operation. Wich one you are trying to solve?

**Comment by Divine1** [comment]
@skabashnyuk 

i have to verify/implement below 2 features..

1. Gitlab as oidc provider for Kubernetes (eclipse che) 
2. the `devfile.yaml` file will be present inside my `gitlab.org.com`. the devworkspace should be able to open the `codeeditor` and clone the source code

**Comment by skabashnyuk** [comment]
ok. can we solve them one by one separately? at which one you are now?

**Comment by Divine1** [comment]
@skabashnyuk currently i'm doing `Gitlab as oidc provider for Kubernetes (eclipse che)`. but i need to know if i can remove Dex static users, delete all enablePasswordDB and staticPasswords sections from `kubectl edit configmap dex -n dex ` configMap

**Comment by skabashnyuk** [comment]
>but i need to know if i have to remove Dex static users, delete all enablePasswordDB and staticPasswords sections from kubectl edit configmap dex -n dex configMap inorder to successfully complete the gitlab as oidc provider integration. 


I think it's a matter of your choice to decide. If you don't want to keep them or something is not working  - you can remove them.

**Comment by Divine1** [comment]
@skabashnyuk 

i have applied the `gitlab-secret` in `eclipse-che` namespace

```
kind: Secret
apiVersion: v1
metadata:
  name: gitlab-oauth-config
  namespace: 'eclipse-che' 
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: oauth-scm-configuration
  annotations:
    che.eclipse.org/oauth-scm-server: gitlab
    che.eclipse.org/scm-server-endpoint: 'https://gitlab.org.com'
type: Opaque
data:
  id: 'ZWFkNDhjNzE1ODQ1ZDU3NDI1M2ExOGRmZTVlOTlkMjhhYjViMmY2OGQ4NDY5ODZmMTQid'
  secret: 'OGM0MDMwNWQ2OWI1ZGQ1ZGQ3ODk0YWQ0ZDc1MjZmMzNjMmZiMTUwMzhiMDJkMGZsecret'
```

then i opened the `eclipse-che` website  `https://192.168.49.2.nip.io`, then logged-in using `che@eclipse.org / admin`

After login, i provided my repository containing valid devfile `https://gitlab.org.com/demoproject.git` -> clicked create button -> the `gitlab.org.com` sso screen appeared -> then i received invalid redirect uri message (attached screenshot below)

![image](https://user-images.githubusercontent.com/10482177/151376820-80042983-94b4-4ea4-bf0d-5c836fafc2ca.png)

![image](https://user-images.githubusercontent.com/10482177/151377216-53502f2d-1571-4850-a9d5-915acb516a12.png)


i have updated below redirect uri, but this throws error. 
i have also tested by adding `https://192.168.49.2.nip.io/callback`. this threw the same `the redirectURI included is not valid` error message.



![image](https://user-images.githubusercontent.com/10482177/151375666-89d5a244-a3a1-4cc6-af7f-aa356ccd387b.png)

**please let me know how to fix this error.**

**Comment by Divine1** [comment]
@skabashnyuk i added `https://192.168.49.2.nip.io/api/oauth/callback` to the redirectURL in `gitlab-outh` configuration under `applications` in `gitlab.org.com`. now the redirection works fine without any issue



i'm facing another issue,


When i remove Dex static users, delete all enablePasswordDB and staticPasswords sections from `kubectl edit configmap dex -n dex configMap` , the `dex` pod under `dex` namespace fails to restart with below error

![image](https://user-images.githubusercontent.com/10482177/151382676-4c1d4081-1b2a-4bb9-a250-ad2505b44d0c.png)


how to replace the default `dex` login screen with `gitlab-sso` based authentication?
![image](https://user-images.githubusercontent.com/10482177/151383063-1c3fb2b5-f6d0-4c66-9c82-9075b6f29f28.png)

**Comment by skabashnyuk** [comment]
>Do i have to replace it with any gitlab connector or something? 

I don't know

**Comment by Divine1** [comment]
@skabashnyuk  thanks a lot for the support, it helped me a lot. i'm looking forward to contribute to this project but not sure in which part i can do.


i will re-open a new ticket for a different issue but related to my above comment.
Have a great day,
